site/content/blog/2020-10-28-docker-content-trust-in-gitlab-with-delegation.md

```diff
@@ -0,0 +1,92 @@
+
+## Create a signed repository and add signer
+
+This is not part of the automation process, because it requires the root key:
+```
+docker trust signer add --key public-key-of-signer.pub signer-name registry/company/repository
+```
+This will ask you for the root key passphrase and needs the encrypted root key locally in `$HOME/.docker/trust/private/ROOT_KEY_HASH.key`.
+This command does two things when first run:
+1. Upgrades the docker repository to use Docker Content Trust and therefore creates a new key, and you need to input a newly generated passphrase (Please back up and version)
+1. Adds the signer named `signer-name` to be allowed to sign new tags pushed
+
+## Automation concept and quirks
+
+- Use the passphrase to decrypt the encrypted keys which are encrypted at rest
+- The `DOCKER_CONTENT_TRUST_REPOSITORY_PASSPHRASE` environment variable to decrypt the key for automation usage is not just for the repository key, it is also for the signer key and root key
+- The keys can only be used one at a time (multiple keys cannot be loaded at the same time, as opposed to e.g. gpg keyring concept)
+- The keys should be located in `$HOME/.docker/trust/private/*` otherwise automated loading of keys fails
+- Should you use `$` or `!` in the passphrases and use them as GitLab CI/CD secret variables, be aware that these characters need to be [escaped](https://stackoverflow.com/questions/48870664/escape-char-in-gitlab-secret-variables)
+
+## Automation of pushing signed image:tag in gitlab
+
+```yaml
+dockerhub-edge:
+  variables:
+    URL: docker.io
+    USERNAME: $DOCKER_HUB_USERNAME # gitlab ci variable type variable
+    TOKEN: $DOCKER_HUB_TOKEN # gitlab ci variable type variable
+    IMAGE: $URL/mattermost/${CI_PROJECT_NAME}
+    TAG: edge
+  before_script:
+    - echo $TOKEN | docker login --username $USERNAME $URL --password-stdin
+  script:
+    - docker build --tag ${IMAGE}:${TAG} .
+    - export DOCKER_CONTENT_TRUST=1
+    - SIGNER_KEY_NAME="CHANGE_TO_SIGNER_KEY_HASH" # change this to your hash
+    - PATH_KEYS=$HOME/.docker/trust/private
+    - mkdir -p $PATH_KEYS
+    - chmod 600 $DCT_SIGNER_PRIV_KEY
+    - cp $DCT_SIGNER_PRIV_KEY $PATH_KEYS/$SIGNER_KEY_NAME.key # gitlab ci variable type file
+
+    - export DOCKER_CONTENT_TRUST_REPOSITORY_PASSPHRASE=$DCT_SIGNER_PASS # gitlab ci variable type variable
+    - docker trust key load $PATH_KEYS/$SIGNER_KEY_NAME.key
+    - docker trust sign ${IMAGE}:${TAG}
+    - docker push ${IMAGE}:${TAG}
+    - docker trust inspect --pretty ${IMAGE}:${TAG}
+  after_script:
+    - docker logout
+    - SIGNER_KEY_NAME="CHANGE_TO_SIGNER_KEY_HASH" # change this to your hash
+    - PATH_KEYS=$HOME/.docker/trust/private
+    - rm $PATH_KEYS/$SIGNER_KEY_NAME.key
+  tags:
+    - docker
+  only:
+    - master
+```
+
+## Advanced image signing features
+
+The CLI command `docker trust` is a wrapper for [notary](https://github.com/theupdateframework/notary).
```

Comment:
I would switch these two sentences around, it might flow better?

Reply:
Thank you for the review, it does indeed flow better! Updated it, if you want to take another look.